text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,339 @@
+                 |               |     |     |     |     
+                 |               |     |     |     |     
+                 |               |     |     |     |     
+                 |   trans_internal ---+   transmogrify   main
+INTERFACES       |       ^    ^                ^     ^       ^
+-----------      |       |    |                |     |       |
+BODIES           |       |    +---------+      |     +----+  |
+                 |       |              |      |          |  | 
+               trans_internal'        transmogrify'       main'
+
+```
+
+Note that, for readability, I've omitted some redundant edges in the above graph. An edge between two nodes can be omitted if there is another path between those nodes (as in `trans_internal' --> Kid`). More formally, two dependency graphs are equivalent for our purposes if their transitive closures are equal.
+
+This dependency graph can be queried for finding all `program item interfaces` that a `program item` `A` depends on: Start at the node corresponding to `A` and collect all transitively reachable nodes. In the above example this means that `main()` needs to be recompiled if `Kid`, `Tiger` or the interface of `transmogrify` change (because these are reachable from `main`'s body node) but not if `Dinosaur`, `Gastropod` or any of the other function's body changes (because their nodes are not reachable from `main`'s body).
+
+## Generic Program Items
+So far the dependency graph has only described non-generic `program items`. For generic definitions the situation is a bit more complicated since their dependencies are only fully defined once all type parameters are substituted for concrete arguments. Consider the following example:
+
+```rust
+trait Transmogrifiable<T> {
+    fn transmogrify(self) -> T;
+}
+
+impl Transmogrifiable<Tiger> for Kid {
+    fn transmogrify(self) -> Tiger { ... }
+}
+
+impl Transmogrifiable<Gastropod> for Dinosaur {
+    fn transmogrify(self) -> Gastropod { ... }
+}
+
+fn transmogrify<TFrom, TTo>(val: TFrom) -> TTo 
+    where TFrom: Transmogrifiable<TTo> 
+{
+    val.transmogrify()        
+}
+```
+
+In this example `transmogrify<Kid, Tiger>` will have a different dependency graph than `transmogrify<Dinosaur, Gastropod>`. In other words, the monomorphized implementation of `transmogrify<Kid, Tiger>` is not affected if the definition of `Dinosaur` or `Gastropod` changes and the dependency graph should reflect this.
+
+One way to model this behavior is by not creating dependency graph nodes for generic `program items` but `node templates`, which---like generic items---have type parameters and yield a concrete, monomorphic dependency graph node, if all type parameters are substituted for concrete arguments. When the need arises to check if a particular monomorphized function implementation from the cache can be re-used, the dependency graph for the function can be constructed on demand from the given `node template` and parameter substitutions.
```

Comment:
If you explicitly stored the list of dependencies of each cache item in the cache, then it wouldn't be necessary to construct anything on demand - whether generic or not, you wouldn't have to go through the function to resolve types and such to determine what it actually depends on, which sounds like an improvement.

In any case, if you're caching object code, you need to store the list of functions that were inlined or otherwise had their behavior consulted by LLVM for codegen of the current function, or else you have to recompile a function whenever any of its transitive dependencies change in any way.